WordPress/src/main/java/org/wordpress/android/ui/sitecreation/verticals/NewSiteCreationVerticalsViewModel.kt

```diff
@@ -0,0 +1,252 @@
+) : ViewModel(), CoroutineScope {
+    private val job = Job()
+    private var fetchVerticalsJob: Job? = null
+    override val coroutineContext: CoroutineContext
+        get() = IO + job
+    private var isStarted = false
+
+    private val _uiState: MutableLiveData<VerticalsUiState> = MutableLiveData()
+    val uiState: LiveData<VerticalsUiState> = _uiState
+
+    private var listState: ListState<VerticalModel> = ListState.Init()
+    private lateinit var headerInfo: DummyVerticalsHeaderInfoModel
+
+    private val _clearBtnClicked = SingleLiveEvent<Void>()
+    val clearBtnClicked = _clearBtnClicked
+
+    init {
+        dispatcher.register(fetchVerticalsUseCase)
+    }
+
+    override fun onCleared() {
+        super.onCleared()
+        dispatcher.unregister(fetchVerticalsUseCase)
+    }
+
+    fun start() {
+        if (isStarted) {
+            return
+        }
+        isStarted = true
+        fetchHeaderInfo()
+    }
+
+    private fun fetchHeaderInfo() {
+        launch {
+            withContext(MAIN) {
+                updateUiStateToFullScreenProgress()
+            }
+            val headerInfoEvent = fetchVerticalsHeaderInfoUseCase.fetchVerticalHeaderInfo()
+            withContext(MAIN) {
+                onHeaderInfoFetched(headerInfoEvent)
+            }
+        }
+    }
+
+    private fun onHeaderInfoFetched(event: DummyOnVerticalsHeaderInfoFetched) {
+        if (event.isError) {
+            updateUiStateToFullScreenError()
+        } else {
+            headerInfo = event.headerInfo!!
+            updateUiStateToContent("", ListState.Ready(emptyList()))
+        }
+    }
+
+    fun onFetchHeaderInfoRetry() {
+        fetchHeaderInfo()
+    }
+
+    fun onClearTextBtnClicked() {
+        _clearBtnClicked.call()
+    }
+
+    fun updateQuery(query: String, delay: Int = throttleDelay) {
+        fetchVerticalsJob?.cancel() // cancel any previous requests
+        if (query.isNotEmpty()) {
+            fetchVerticals(query, delay)
+        } else {
+            updateUiStateToContent(query, ListState.Ready(emptyList()))
+        }
+    }
+
+    private fun fetchVerticals(query: String, throttleDelay: Int) {
+        fetchVerticalsJob = launch {
+            withContext(MAIN) {
+                updateUiStateToContent(query, ListState.Loading(ListState.Ready(emptyList()), false))
+            }
+            delay(throttleDelay)
+            val fetchedVerticals = fetchVerticalsUseCase.fetchVerticals(query)
+            withContext(MAIN) {
+                onVerticalsFetched(query, fetchedVerticals)
+            }
+        }
+    }
+
+    private fun onVerticalsFetched(query: String, event: OnVerticalsFetched) {
+        if (event.isError) {
+            updateUiStateToContent(query, ListState.Error(listState, event.error.message))
+        } else {
+            updateUiStateToContent(query, ListState.Success(event.verticalList))
+        }
+    }
+
+    private fun updateUiStateToFullScreenProgress() {
```

Comment:
I'd prefer to combine `updateUiStateToFullScreenProgress` and `updateUiStateToFullScreenError` with a parameter as `showContent`, `showSkipButton` and `items` should be the same for both, and `showFullscreenError` and `showFullscreenProgress` would be exact opposites. I think it'd make it easier to understand this code.

Reply:
I wanted to do that as well, but I couldn't think of any good name for neither the method nor the parameter. So this solution looked like it'd bring more clarity. I'll be happy to change it if we can come up with a reasonable names. Any suggestions?